### DISCUSSION TITLE
SQL上线检测和提交

### DISCUSSION THREAD

**Discussion by peixubin** [hhyo/Archery#2691] [Q&A] [answered]
在SQL上线检测时出现错误的情况，是否应该不允许提交，至少应该有个开关控制，不允许强制提交。

**Comment by LeoQuote** [comment]
有相关配置，可以探索下

**Reply by peixubin** [reply]
具体是哪个配置项？ 我在《配置管理->SQL上线》部分看了，没有合适的

**Reply by peixubin** [reply, marked as answer]
找到了问题原因。
在AuditV2类中，把自动通过和自动审核等价了。这应该是个**严重BUG**了，希望能尽快修复。
`
    def generate_audit_setting(self) -> AuditSetting:
        if self.is_auto_review():
            return AuditSetting(auto_pass=True)
        ...
    def create_audit(self):
        ...
        audit_setting = self.generate_audit_setting()
        ...
        # 自动通过的情况
        if audit_setting.auto_pass:
            self.audit.current_status = WorkflowStatus.PASSED
`

**Reply by LeoQuote** [reply]
我归纳下, 当前数据库如果设置了"自动审批", 也就是 "AUTO_REVIEW", 那么只要提交的语句不命中设置中的正则, 即使规则引擎判定该语句不能提交, 这个语句也是可以提交成功的.

这确实是实现上的问题, 应该在 https://github.com/hhyo/Archery/blob/24f86de92951cd762055bb61d15c55e04a6164b8/sql/utils/workflow_audit.py#L258 这个方法里, 尽快判断当前的workflow 状态, 如果是 autoreviewwrong, 则直接 raise exception.

欢迎 pr 哈. 我暂时没空修这个

**Reply by peixubin** [reply]
参考 #2692，看上去是修复了

**Comment by peixubin** [comment]
只看到critical_ddl_regex和auto_review_wrong像是实现类似的功能。
critical_ddl_regex是通过正则来匹配语句，和我的需求不一致；
auto_review_wrong看上去像，但是我配置成2,不起作用

**Reply by lanheader** [reply]
https://hanchuanchuan.github.io/goInception/zh/rules.html#ddl%E6%93%8D%E4%BD%9C  看看goinception的配置

**Reply by peixubin** [reply]
上线脚本已经通过goinception发现了错误，在SQL提交时会提示“存在N个错误”，让用户确认。用户确认后，存在错误的脚本就会提交到审核（有些还会自动审核通过）。
我的意思是存在错误的脚本，不能提交到审核或自动审核通过
![截图 2024-06-23 10-22-44](https://github.com/hhyo/Archery/assets/23612384/2fe0e863-ccfe-4744-bb46-ffefe48ae668)
![截图 2024-06-23 10-23-08](https://github.com/hhyo/Archery/assets/23612384/875c7dca-ae34-4230-8225-3ef53e7dfb52)
。

**Comment by peixubin** [comment]
想要达到的效果：
![截图 2024-06-23 10-45-55](https://github.com/hhyo/Archery/assets/23612384/93c8074e-4bcf-4fa9-ade8-defce7ecffa1)

**Reply by feiazifeiazi** [reply]
强制提交了，有什么关系？ 会自动审核不通过的。

你这里为什么会自动审核通过，我擦，咋通过的？

**Reply by peixubin** [reply]
我也很奇怪，SQL上线相关配置项：
![auto_review_conf](https://github.com/hhyo/Archery/assets/23612384/487a7e09-7373-473b-b363-8e1cea0d9462)
是开启了自动审核的。
如果我在AUTO_REVIEW_REGEX加入"^insert"，就是把insert语句需要人工审核来处理，那么提交后工作流状态将是“自动审核不通过”，是正确的。但是这样的话，所有的insert都会进入人工审核环节，这不符合我的业务需求，我的目的是DDL或更新行数比较多的要人工审核。

**Comment by gppwin** [comment]
大佬最后怎么解决的，系统有设置的地方吗？
